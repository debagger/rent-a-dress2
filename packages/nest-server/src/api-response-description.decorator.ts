import { ApiResponseMetadata } from '@nestjs/swagger';

export const ApiResponseDescription = (
  statusCode: string,
  description: string,
): MethodDecorator => {
  const decoratorFactory = (target: object, key?: any, descriptor?: any) => {
    const responses =
      Reflect.getMetadata('swagger/apiResponse', descriptor.value) || {};
    const response: ApiResponseMetadata = responses[statusCode] || {};

    response.description = response.description || description;
    Reflect.defineMetadata('swagger/apiResponse', responses, descriptor.value);

    return descriptor;
  };
  decoratorFactory.KEY = 'swagger/apiResponse';
  return decoratorFactory;
};
