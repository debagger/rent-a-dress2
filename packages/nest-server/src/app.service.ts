import * as entities from '@rent-a-dress/entities';
import { Injectable, Inject } from '@nestjs/common';
import { Connection, Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { User } from '@rent-a-dress/entities';

@Injectable()
export class AppService {
  constructor(@InjectRepository(User) private readonly repo: Repository<User>) {}

  async getHello(): Promise<string> {
    const users = this.repo;
    const newUser = await users.save({
      username: 'Admin',
      email: 'admin@rent-a-dress.ru',
      role: 'admin',
      password: 'pass',
    });
    console.log(`User id=${newUser.id} was created`);
    return `${await users.count()} users `;
  }
}
