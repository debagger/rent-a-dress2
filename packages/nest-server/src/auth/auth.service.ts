import { Injectable } from '@nestjs/common';
import { UsersService } from '../Users/users.service';
import { createHash } from 'crypto';
import { JwtService } from '@nestjs/jwt';

@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
  ) {}

  async validateUser(
    userName: string,
    password: string,
  ): Promise<
    { id: number; username: string; email: string; role: string } | undefined
  > {
    const user = await this.usersService.findOne({ username: userName });
    if (user && user.password === this.usersService.hash(password)) {
      const { password, tokens, ...result } = user;
      return result;
    }
    return;
  }

  async login(user: any) {
    const payload = { username: user.username, sub: user.id };
    return {
      access_token: this.jwtService.sign(payload),
    };
  }
}
