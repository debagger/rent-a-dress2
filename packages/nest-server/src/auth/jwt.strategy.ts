import { ExtractJwt, Strategy } from 'passport-jwt';
import { PassportStrategy } from '@nestjs/passport';
import { Injectable } from '@nestjs/common';
import { jwtConstants } from './constants';
import { UsersService } from '../users/users.service';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor(private readonly users: UsersService) {
    super({
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
      ignoreExpiration: true,
      secretOrKey: jwtConstants.secret,
    });
  }

  async validate(payload: any) {
    console.log("JwtStrategy payload = ", payload);
    const dbUser = await this.users.findOne(payload.sub);
    const {password, ...user} = dbUser
    return { payload, user };
  }
}