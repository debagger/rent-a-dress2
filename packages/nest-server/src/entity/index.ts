export * from "./image.entity"
export * from "./token.entity"
export * from "./user.entity"
export * from "./user.interface"
export * from "./catalog-item-option.entity"
export * from "./catalog-item.entity"