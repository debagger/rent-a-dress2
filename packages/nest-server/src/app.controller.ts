import { Controller, Get, UseGuards, Post, Request } from '@nestjs/common';
import { AppService } from './app.service';
import { AuthGuard } from '@nestjs/passport';
import { AuthService } from './auth/auth.service';
@Controller()
export class AppController {
  constructor(
    private readonly appService: AppService,
    private readonly authService: AuthService,
  ) {}

  @Get()
  async getHello(): Promise<string> {
    return this.appService.getHello();
  }

  @UseGuards(AuthGuard('local'))
  @Post('auth/login')
  async login(@Request() req): Promise<{ access_token: String }> {
    return this.authService.login(req.user);
  }

  @UseGuards(AuthGuard('jwt'))
  @Get('profile')
  async getProfile(@Request() req) {
    console.log(req.user);
    return req.user;
  }
}
