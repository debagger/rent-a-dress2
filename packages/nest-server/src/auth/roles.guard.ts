import { Injectable, ExecutionContext, CanActivate } from '@nestjs/common';
import { AuthGuard, IAuthGuard, Type } from '@nestjs/passport';
import { Observable } from 'rxjs';
import { AuthModule } from './auth.module';
import { bootstrap } from 'src/main';
import { Reflector } from '@nestjs/core';

@Injectable()
export class RoleGuard extends AuthGuard('jwt') {
  constructor(private readonly reflector: Reflector) {
    super();
  }

  canActivate(context: ExecutionContext) {
    const roles = this.reflector.get<string[]>('roles', context.getHandler());
    if (!roles) return true;
    const result = super.canActivate(context);
    if (roles.length == 0) return result;
    if (result instanceof Promise) {
      return (<Promise<boolean>>result).then(res => {
        if (res) {
          const userRole = context.switchToHttp().getRequest().user?.user?.role;
          const isUserInRole = !!roles.find(item => item === userRole);
          return isUserInRole;
        }
      });
    }
    return result;
  }
}
