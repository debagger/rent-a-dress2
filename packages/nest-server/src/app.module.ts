import { Module, NestModule, MiddlewareConsumer } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { UsersModule } from './users/users.module';
import { TypeOrmModule, InjectRepository } from '@nestjs/typeorm';
import { AuthModule } from './auth/auth.module';
import * as entities from './entity';
import { Repository } from 'typeorm';
import { User } from './entity';
import { UsersService } from './users/users.service';
import { CatalogModule } from './catalog/catalog.module';

@Module({
  imports: [
    UsersModule,
    TypeOrmModule.forRoot({
      type: 'sqlite',
      database: ':memory:',
      entities: Object.values(entities),
      synchronize: true,
      keepConnectionAlive: true,
    }),
    TypeOrmModule.forFeature(Object.values(entities)),
    AuthModule,
    CatalogModule,
  ],
  controllers: [AppController],
  providers: [AppService],
  exports:[AuthModule, CatalogModule]
})
export class AppModule implements NestModule {
  constructor(private readonly users: UsersService) {}
  async configure() {
    let admin = await this.users.findOne({ username: 'admin' });
    if (!admin) {
      admin = await this.users.save({
        username: 'admin',
        email: 'debagger@gmail.com',
        role: 'admin',
        password: '123',
      });
    }
  }
}
