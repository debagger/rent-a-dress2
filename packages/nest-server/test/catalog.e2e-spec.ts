import { Test } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import { AppModule } from '../src/app.module';
import { getRepository, DeepPartial } from 'typeorm';
import { catalogItem, User } from '../src/entity';
import * as request from 'supertest';
import { getAdminToken } from './getAdminToken';
import { UsersService } from '../src/users/users.service';
import { AuthService } from '../src/auth/auth.service';
import { AllExceptionsFilter } from '../src/exception.filter';
import { type } from 'os';

describe('Catalog API tests (e2e)', () => {
  let app: INestApplication;
  let userToken: string;
  beforeEach(async () => {
    const ModuleFixture = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();
    app = ModuleFixture.createNestApplication();
    app.useGlobalFilters(new AllExceptionsFilter());
    app.useGlobalPipes(new ValidationPipe());
    await app.init();

    const userService = app.get(UsersService);
    const authService = app.get(AuthService);

    const user = await userService.save(<DeepPartial<User>>{
      username: 'user',
      password: '123',
      role: 'user',
      email: '',
    });

    userToken = (await authService.login(user)).access_token;

    const db = getRepository(catalogItem);
    const items = [...Array(10).keys()].map(index => {
      return <Partial<catalogItem>>{
        caption: `Item ${index}`,
        desc: `Item description ${index}`,
        img: '',
        price: index * 1000,
      };
    });
    await db.save(items);
  });

  afterEach(async () => {
    await getRepository(catalogItem).clear();
    await getRepository(User).delete({ username: 'user' });
  });

  it('/catalog (GET)', async () => {
    const server = app.getHttpServer();
    return request(server)
      .get('/catalog')
      .expect(200)
      .expect(res => {
        expect(res.body.total).toBe(10);
        expect(res.body.data).toMatchSnapshot();
      });
  });

  it('/catalog (POST)', async () => {
    const server = app.getHttpServer();
    const token = await getAdminToken(server);
    return request(server)
      .post('/catalog')
      .auth(token, { type: 'bearer' })
      .send(<Partial<catalogItem>>{
        caption: 'New CatalogItem',
        img: '',
        desc: 'New CatalogItem description',
        price: 1000,
      })
      .expect(201, {
        id: 21,
        caption: 'New CatalogItem',
        img: '',
        desc: 'New CatalogItem description',
        price: 1000,
      });
  });

  it('/catalog (POST) using user token', async () => {
    const server = app.getHttpServer();

    return request(server)
      .post('/catalog')
      .auth(userToken, { type: 'bearer' })
      .send(<Partial<catalogItem>>{
        caption: 'New CatalogItem',
        img: '',
        desc: 'New CatalogItem description',
        price: 1000,
      })
      .expect(403);
  });

  it('/catalog (POST) no token', async () => {
    const server = app.getHttpServer();

    return request(server)
      .post('/catalog')
      .send(<Partial<catalogItem>>{
        caption: 'New CatalogItem',
        img: '',
        desc: 'New CatalogItem description',
        price: 1000,
      })
      .expect(401);
  });

  it('/catalog (POST) check validation', async () => {
    const server = app.getHttpServer();
    const token = await getAdminToken(server);

    return request(server)
      .post('/catalog')
      .auth(token, { type: 'bearer' })
      .send({
        caption: 'New CatalogItem',
        img: '',
        desc: 'New CatalogItem description',
        price: '1000 руб',
      })
      .expect(res => {
        expect(res.status).toBe(400);
        expect(res.body).toMatchSnapshot();
      });
  });
});
