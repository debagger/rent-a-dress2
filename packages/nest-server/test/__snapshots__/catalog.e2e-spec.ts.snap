// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog API tests (e2e) /catalog (GET) 1`] = `
Array [
  Object {
    "caption": "Item 0",
    "desc": "Item description 0",
    "id": 1,
    "img": "",
    "price": 0,
  },
  Object {
    "caption": "Item 1",
    "desc": "Item description 1",
    "id": 2,
    "img": "",
    "price": 1000,
  },
  Object {
    "caption": "Item 2",
    "desc": "Item description 2",
    "id": 3,
    "img": "",
    "price": 2000,
  },
  Object {
    "caption": "Item 3",
    "desc": "Item description 3",
    "id": 4,
    "img": "",
    "price": 3000,
  },
  Object {
    "caption": "Item 4",
    "desc": "Item description 4",
    "id": 5,
    "img": "",
    "price": 4000,
  },
  Object {
    "caption": "Item 5",
    "desc": "Item description 5",
    "id": 6,
    "img": "",
    "price": 5000,
  },
  Object {
    "caption": "Item 6",
    "desc": "Item description 6",
    "id": 7,
    "img": "",
    "price": 6000,
  },
  Object {
    "caption": "Item 7",
    "desc": "Item description 7",
    "id": 8,
    "img": "",
    "price": 7000,
  },
  Object {
    "caption": "Item 8",
    "desc": "Item description 8",
    "id": 9,
    "img": "",
    "price": 8000,
  },
  Object {
    "caption": "Item 9",
    "desc": "Item description 9",
    "id": 10,
    "img": "",
    "price": 9000,
  },
]
`;

exports[`Catalog API tests (e2e) /catalog (POST) check validation 1`] = `
Object {
  "path": "/catalog",
  "statusCode": 400,
}
`;
